Detectors/Base/include/DetectorsBase/Track.h

```diff
@@ -63,158 +70,140 @@ namespace AliceO2 {
+          float GetSigmaTglSnp()               const { return mC[kSigTglSnp]; }
+          float GetSigmaTgl2()                 const { return mC[kSigTgl2]; }
+          float GetSigma1PtY()                 const { return mC[kSigQ2PtY]; }
+          float GetSigma1PtZ()                 const { return mC[kSigQ2PtZ]; }
+          float GetSigma1PtSnp()               const { return mC[kSigQ2PtSnp]; }
+          float GetSigma1PtTgl()               const { return mC[kSigQ2PtTgl]; }
+          float GetSigma1Pt2()                 const { return mC[kSigQ2Pt2]; }
+
+          void  Print()                        const;
+
+          // parameters + covmat manipulation
           bool  Rotate(float alpha);
           bool  PropagateTo(float xk, float b);
-          bool  PropagateTo(float xk, const float b[3]);
+          bool  PropagateTo(float xk, const array<float,3> &b);
           void  Invert();
 
-          float GetPredictedChi2(const float p[2], const float cov[3]) const;
-          bool  Update(const float p[2], const float cov[3]);
+          float GetPredictedChi2(const array<float,2> &p, const array<float,3> &cov) const;
+          bool  Update(const array<float,2> &p, const array<float,3> &cov);
 
           bool  CorrectForMaterial(float x2x0,float xrho,float mass,bool anglecorr=false,float dedx=kCalcdEdxAuto);
 
           void  ResetCovariance(float s2=0);
           void  CheckCovariance();
-          void  Print();
 
         protected:
-          // internal cast to TrackPar
-          const TrackPar* Param()              const { return reinterpret_cast<const TrackPar*>(this); }
-          TrackPar* Param()                          { return reinterpret_cast<TrackPar*>(this); }
-          bool TrackPar2Momentum(float p[3], float alpha);
+          float mC[kCovMatSize] = {0.f};  // 15 covariance matrix elements
 
-        protected:
-          float mPC[kTrackPCSize];  // x, alpha + 5 parameters + 15 errors
-
-          static const float kCalcdEdxAuto; // value indicating request for dedx calculation
       };
 
+      class TrackPar final : public TrackParBase { // track parameterization only
+        public:
+          TrackPar() {}
+          TrackPar(float x,float alpha, const array<float,kNParams> &par) : TrackParBase{x,alpha,par} {}
+          TrackPar(const array<float,3> &xyz, const array<float,3> &pxpypz,int sign, bool sectorAlpha=true);
+          //
+          void  Print() const {PrintParam();}
+      };
 
       //____________________________________________________________
-      inline TrackPar::TrackPar(float x, float alpha, const float par[kNParams]) {
+      inline TrackParBase::TrackParBase(float x, float alpha, const array<float, kNParams> &par) : mX{x}, mAlpha{alpha} {
         // explicit constructor
-        mP[kX] = x;
-        mP[kAlpha] = alpha;
-        memcpy(&mP[kY],par,kNParams*sizeof(float));
+        std::copy(par.begin(), par.end(), mP);
```

Comment:
@ktf, in principle I agree on the second point, if you mean a point in space, express it in the code, but please don't use a ROOT type for that :) Also, simple std:: types have the big advantage of being, well, standard and performant (as opposed to obscure; easier to look up). If you have a point type somewhere and you need to put it in this interface, well, make sure it decays to std::array in my opinion.

as for std::array - as I mentioned, the standard guarantees same semantics - which means API, there are no guarantees as to the layout. There might be padding, some alignment, all kinds of stuff that breaks the layout without breaking the spec. For classes that are meant to be flat we should express that directly by using (carrays of) POD.